### DISCUSSION TITLE
fastapi build docker and publish to  k8s and i need gunicorn+fastapi or not ?

### DISCUSSION THREAD

**Discussion by un-human** [fastapi/fastapi#7178] [Questions]
Linux environment
Question 1：fastapi build docker and publishto  k8s, so do  i need gunicorn+fastapi or not ?? , just uvicorn+fastapi ?
gunicorn+fastapi  -w 1   pk  uvicorn+fastapi,  Which performance is higher ??

Question 2：Can I release directly to the production environment with  uvicorn+fastapi （no gunicorn and build docker to k8s）??

tks vm!!

**Comment by clmnin** [comment]
Uvicorn's doc for deployment says ([Source](https://www.uvicorn.org/deployment/))
1. Run `uvcorn --reload` from the terminal for local development
2. Run `gunicorn -k uvicorn.workers.UvicornWorker` for production.
3. Additionally run behind Nginx for self-hosted deployments.
4. Finally, run everything behind a CDN for caching support, and serious DDOS protection.

**gunicorn has a pre-fork worker model**

A pre-fork worker model basically means a master creates forks which handle each request. A fork is a completely separate *nix process ([Source](https://stackoverflow.com/questions/25834333/what-exactly-is-a-pre-fork-web-server-model)). And uvicorn is a ASGI server running [uvloop](https://github.com/MagicStack/uvloop) (every python async needs an event loop and you use uvloop here)

So that means gunicorn in front of uvcorn will help you have workers handle requests before it hits uvcorn which itself is a ASGI server. So for scalability you should be using `gunicorn -k uvicorn.workers.UvicornWorker`. Also please go though the uvcorn deployment doc mentioned above.

**Comment by clmnin** [comment]
To answer your questions
1. Which performance is higher?
**A.** gunicorn on top of uvicorn will perform better (because you have multiple workers/*nix processes which is scheduled by the OS for performance and concurrency)

2. Can I release directly to the production environment with uvicorn+fastapi
**A.** Yes you can

**Comment by un-human** [comment]
> 1. gunicorn on top of uvicorn will perform better (because you have multiple workers/*nix processes which is scheduled by the OS for performance and concurrency)

k8s can create multiple Pods to implementation these concurrent processes Instead of Gunicorn
so use Gunicorn to manage Uvicorn and run multiple of these concurrent processes can forget

i mean:  gunicorn+fastapi -w 1 pk uvicorn+fastapi, Which performance is higher ?? 
attention this: gunicorn -w 1 
so Question:
gunicorn+fastapi -w 1（1 worker）   pk    uvicorn+fastapi(no gunicorn manage ), Which performance is higher ??

**Comment by clmnin** [comment]
I'm sorry I didn't quite catch the question the first time around, now amusing `pk` means `vs` here is my answer:

These technologies (gunicorn and k8) aren't as similar as they initially seem. They address different portions of the application stack and are actually complementary.

### Gunicorn
Web request concurrency is primarily limited by network I/O or "I/O bound". These types of tasks can be scaled using cooperative scheduling provided by threads. If you find request concurrency is limiting your application, increasing gunicorn worker threads may well be the place to start

### Kubernetes
Where Kubernetes comes in handy is by providing out-of-the-box horizontal scalability and fault tolerance.

### Uvicorn
Uvicorn is a lightning-fast "ASGI" server. It runs asynchronous Python web code in a **single process**.

So for `gunicorn + uvicorn + fastapi` with one worker deployment what you get is
1. One async worker (like a Gevent that creates a new greenlet for every new request). 
    * Every time a new request comes they are handled by greenlets spawned by the worker threads. The core idea is when there is IO bound action in your code the greenlet will yield and give a handle to another greenlet to use the CPU. So when many requests are coming to your app, it can handle them concurrently, the IO-bound calls are not going to block the CPU and throughput will improve.
2. `uvloop` as your async loop
3. `fastapi` as your code

So for `uvicorn + fastapi` with one worker deployment (no gunicorn) what you get is
1. `uvloop` as your async loop
2. `fastapi` as your code

Here you don't have `greenlets` that take care of IO bound tasks and give you true concurrency.

**Using k8 is just an added stack and doesn't replace your gunicorn**


Answer to question:
**Q** gunicorn+fastapi -w 1（1 worker） pk uvicorn+fastapi(no gunicorn manage ), Which performance is higher ??
**A** `gunicorn+fastapi -w 1` will be performant (note: you should be using uvcorn as it uses uvloop, which is faster than asyncio's implementation)

**Comment by un-human** [comment]
> I'm sorry I didn't quite catch the question the first time around, now amusing `pk` means `vs` here is my answer:
> 
> These technologies (gunicorn and k8) aren't as similar as they initially seem. They address different portions of the application stack and are actually complementary.
> 
> ### Gunicorn
> Web request concurrency is primarily limited by network I/O or "I/O bound". These types of tasks can be scaled using cooperative scheduling provided by threads. If you find request concurrency is limiting your application, increasing gunicorn worker threads may well be the place to start
> 
> ### Kubernetes
> Where Kubernetes comes in handy is by providing out-of-the-box horizontal scalability and fault tolerance.
> 
> ### Uvicorn
> Uvicorn is a lightning-fast "ASGI" server. It runs asynchronous Python web code in a **single process**.
> 
> So for `gunicorn + uvicorn + fastapi` with one worker deployment what you get is
> 
> 1. One async worker (like a Gevent that creates a new greenlet for every new request).
>    
>    * Every time a new request comes they are handled by greenlets spawned by the worker threads. The core idea is when there is IO bound action in your code the greenlet will yield and give a handle to another greenlet to use the CPU. So when many requests are coming to your app, it can handle them concurrently, the IO-bound calls are not going to block the CPU and throughput will improve.
> 2. `uvloop` as your async loop
> 3. `fastapi` as your code
> 
> So for `uvicorn + fastapi` with one worker deployment (no gunicorn) what you get is
> 
> 1. `uvloop` as your async loop
> 2. `fastapi` as your code
> 
> Here you don't have `greenlets` that take care of IO bound tasks and give you true concurrency.
> 
> **Using k8 is just an added stack and doesn't replace your gunicorn**
> 
> Answer to question:
> **Q** gunicorn+fastapi -w 1（1 worker） pk uvicorn+fastapi(no gunicorn manage ), Which performance is higher ??
> **A** `gunicorn+fastapi -w 1` will be performant (note: you should be using uvcorn as it uses uvloop, which is faster than asyncio's implementation)


Thank you very much！！！，Thank you

I got it.. use gunicorn+fastapi( use gunicorn manage)  is faster and  performant  ,such as:
  gunicorn fastapi:app -b 0.0.0.0:port -w 2 -k uvicorn.workers.UvicornWorker
then 
  build docker  and publish to  k8s,
（Even create 2 fastapi-Pods，Each container need gunicorn -w 2 -k uvicorn.workers.UvicornWorker ,
 it is faster and  performant !!
）

k8s create 2 fastapi-Pods  doesn't replace  " gunicorn  -w 2  -k uvicorn.workers.UvicornWorker "；its different

Thank you

**Comment by clmnin** [comment]
Please close the issue if this answers your question. Also don't forget to close tiangolo/uvicorn-gunicorn-fastapi-docker#64

**Comment by Gatsby-Lee** [comment]
@stringk245 
I think your question is still valid.
If I rephrase your question, the question becomes which performs between these two
1. a Pod running "uvicorn + FastAPI"
2. a Pod running "gunicorn + unicorn + FastAPI"

And, I think the answer of @clmno is not quite right.

**Comment by mauermbq** [comment]
why not using K8S autoscaler? https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/
see https://github.com/tiangolo/fastapi/issues/1484

**Comment by Gatsby-Lee** [comment]
@mauermbq  
You're right why not just k8s autoscaler. I agree with you.
I think it is just how we want to manage it.

I guess there might be one who want to have a Pod running N workers + HPA to run more Pods
I think it doesn't have to be binary question. It can be optional.

Oh. and I saw something on stackoverflow. Someone did benchmark testing for these
1. Gunicorn ( Uvicorn 4 workers ): Gunicorn does management.
2. Uvicorn 4 workers: Uvicorn does management.

The benchmark result shows that Gunicorn performs better.

For myself, I would use Uvicorn itself if I run one Uvicorn worker in a Pod.
If I use more than one Uvicorn workers in a Pod, I would use Gunicorn.

**Comment by Kludex** [comment]
> The benchmark result shows that Gunicorn performs better.


Do you have the source?

**Comment by un-human** [comment]
> @stringk245
> I think your question is still valid.
> If I rephrase your question, the question becomes which performs between these two
> 
> 1. a Pod running "uvicorn + FastAPI"
> 2. a Pod running "gunicorn + unicorn + FastAPI"
> 
> And, I think the answer of @clmno is not quite right.

yes, That's the question, 3Q
 1. a Pod running "uvicorn + FastAPI" 
 2. a Pod running "gunicorn + unicorn + FastAPI"
 question：Which performance is higher 
answer:, i use gunicorn + unicorn + FastAPI

**Comment by antonwnk** [comment]
@Gatsby-Lee Could you maybe clarify what seems wrong in the "accepted answer"? It's all quite confusing for beginners. I, for one, don't really understand what the gunicorn greenlets do for us when using fastapi and what is the interplay between these and the uvloop?. Maybe @Kludex has some light to shed over this? 😀

**Comment by Gatsby-Lee** [comment]
> > The benchmark result shows that Gunicorn performs better.
> 
> Do you have the source?

I this this is the one 
https://stackoverflow.com/questions/66254371/fastapi-gunicorn-uvicorn-for-production-deployment-with-google-cloud-run-stress

**Comment by Gatsby-Lee** [comment]
@antonwnk  I am not sure which part makes you confused.
- gunicorn: this just manage unicorn workers ( that's it!! )
- uvicorn: ASGI application server ( this also has an option to create N workers )

**Comment by Kludex** [comment]
> > > The benchmark result shows that Gunicorn performs better.
> > 
> > 
> > Do you have the source?
> 
> I this this is the one
> https://stackoverflow.com/questions/66254371/fastapi-gunicorn-uvicorn-for-production-deployment-with-google-cloud-run-stress

@ycd Would you mind sharing the code you used to run the benchmark? (He is the author of the benchmarks)

I've performed a benchmark comparing those technologies, and I didn't get those results.

**Comment by Kludex** [comment]
> @Gatsby-Lee Could you maybe clarify what seems wrong in the "accepted answer"? It's all quite confusing for beginners. I, for one, don't really understand what the gunicorn greenlets do for us when using fastapi and what is the interplay between these and the uvloop?. Maybe @Kludex has some light to shed over this?

There is no greenlet involved on gunicorn with uvicorn. You can check the way the workers are spawned [here](https://github.com/benoitc/gunicorn/blob/ff58e0c6da83d5520916bc4cc109a529258d76e1/gunicorn/arbiter.py#L567). 

Gunicorn only manages the workers, in the sense that it will make sure that if you wish to use 4 workers, you'll always have 4 workers (if they die, they are resurrected) . Uvicorn doesn't do that for you, it only spawns the workers, and in case some worker die, it doesn't care.

Giving the paragraph above, you don't **need** Gunicorn for K8s. You can use it, but it's not really needed. If you're not using it, and you are using multiple uvicorn workers, be sure to have a health condition that makes sense for them (considering my previous paragraph).

uvloop is just an event loop implementation. It's faster than the standard `asyncio`, hence it's recommended for production.

**Comment by antonwnk** [comment]
Perfect, thanks! This clears things up for me.

**Comment by un-human** [comment]
###  manager.py
```python

import asyncio
import platform
import uvicorn

if platform.system() != 'Windows':
    import uvloop

    asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())

from fastapi import FastAPI
from fastapi.responses import ORJSONResponse

app = FastAPI(title='fastapi-service', default_response_class=ORJSONResponse, openapi_url=None, docs_url=None,
              redoc_url=None, swagger_ui_oauth2_redirect_url=None)


@app.post('/index')
async def index():
    return {'name': 'a big dict as response'}


if __name__ == '__main__':
    uvicorn.run(app, host='0.0.0.0', port=9000, debug=False)

```

###  uvicorn  VS  gunicorn
```bash
uvicorn manager:app --host 0.0.0.0 --port 9000 --workers 4

# VS

gunicorn manager:app  -b 0.0.0.0:9001 -w 4 -k uvicorn.workers.UvicornWorker

# result：gunicorn -k uvicorn.workers.UvicornWorker  has higher performance ！！！！
# result2:  advice Gunicorn for K8s

```

### Another 
- url：https://www.techempower.com/benchmarks/#section=data-r20&hw=ph&test=composite
- higher performance:  starlette  >  sanic  > fastapi
- fastapi = starlette  + pydantic ( but  sanic  > fastapi )
- advice:   starlette + pydantic
